driver/src/price_estimation/threaded_price_source.rs

```diff
@@ -0,0 +1,197 @@
+pub struct ThreadedPriceSource {
+    // Shared between this struct and the thread. The background thread writes
+    // the prices, the thread calling `get_prices` reads them.
+    non_blocking_price_source: Arc<Mutex<ManuallyUpdatedPriceSource>>,
+    // Channel over which the thread can be told to update the prices.
+    channel_to_thread: Sender<ThreadMessage>,
+}
+
+impl ThreadedPriceSource {
+    /// The join handle represents the background thread. It can be used to
+    /// verify that it exits when the created struct is dropped.
+    #[allow(dead_code)]
+    pub fn new<T: 'static + PriceSource + Send>(
+        price_source: T,
+        update_interval: Duration,
+    ) -> (Self, JoinHandle<()>) {
+        let non_blocking_price_source = Arc::new(Mutex::new(ManuallyUpdatedPriceSource::new()));
+        let (sender, receiver) = mpsc::channel::<ThreadMessage>();
+
+        let nbps = non_blocking_price_source.clone();
+        // vk: Clippy notes that the following loop and match can be written as
+        // a `while let` loop but I find it clearer to make all cases explicit.
+        #[allow(clippy::while_let_loop)]
+        let join_handle = thread::spawn(move || loop {
+            match receiver.recv_timeout(update_interval) {
+                Ok(ThreadMessage::UpdateImmediately) | Err(RecvTimeoutError::Timeout) => {
+                    let now = Instant::now();
+                    let cutoff = now.checked_sub(update_interval).unwrap_or(now);
+                    let tokens = nbps.lock().unwrap().tokens_that_need_updating(cutoff);
+                    if tokens.is_empty() {
+                        continue;
+                    }
+                    // Make sure we don't hold the mutex while this blocking call happens.
+                    match price_source.get_prices(&tokens) {
+                        Ok(prices) => nbps.lock().unwrap().update_tokens(&tokens, &prices, now),
+                        Err(err) => log::warn!("price_source::get_prices failed: {}", err),
+                    }
+                }
+                // The owning struct has been dropped.
+                Err(RecvTimeoutError::Disconnected) => break,
+            }
+        });
+
+        (
+            Self {
+                non_blocking_price_source,
+                channel_to_thread: sender,
+            },
+            join_handle,
+        )
+    }
+}
+
+impl PriceSource for ThreadedPriceSource {
+    /// Non blocking.
+    /// Infallible.
+    fn get_prices(&self, tokens: &[Token]) -> Result<HashMap<TokenId, u128>> {
+        let mut nbps = self.non_blocking_price_source.lock().unwrap();
+        nbps.track_tokens(tokens);
+        // At this point tokens that were not being tracked already have been
+        // added but they do not yet have prices. This is expected because this
+        // function is supposed to be non blocking.
+        let result = nbps.get_prices(tokens);
+        std::mem::drop(nbps);
```

Comment:
I accidentally deleted your question what the difference between a new block and manually dropping is.

There is no functional difference. A block is more obvious but forces indentation and if you have intermediate variables created in the block that you want to keep outside you have to return them.